test/jest/playwrightEnvironment.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright Microsoft Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const { Playwright } = require('../../lib/server/playwright');
+const NodeEnvironment = require('jest-environment-node');
+const path = require('path');
+const playwright = new Playwright(__dirname, require(path.join(__dirname, '..', '..', 'browsers.json'))['browsers']);
+
+class PlaywrightEnvironment extends NodeEnvironment {
+  constructor(config, context) {
+    super(config, context);
+    this.fixturePool = new FixturePool();
+
+    this.global.registerFixture = (name, fn) => {
+      this.fixturePool.registerFixture(name, 'test', fn);
+    };
+
+    this.global.registerWorkerFixture = (name, fn) => {
+      this.fixturePool.registerFixture(name, 'worker', fn);
+    };
+
+    this.global.registerWorkerFixture('browser', async (test) => {
+      const browser = await playwright.chromium.launch();
+      await test(browser);
+      await browser.close();
+    });
+
+    this.global.registerFixture('context', async (browser, test) => {
+      const context = await browser.newContext();
+      await test(context);
+      await context.close();
+    });
+
+    this.global.registerFixture('page', async (context, test) => {
+      const page = await context.newPage();
+      await test(page);
+    });
+  }
+
+  async setup() {
+    await super.setup();
+  }
+
+  async teardown() {
+    await this.fixturePool.teardownScope('worker');
+    await super.teardown();
+  }
+
+  runScript(script) {
+    return super.runScript(script);
+  }
+
+  async handleTestEvent(event, state) {
+    if (event.name === 'test_start' && event.test) {
+      const fn = event.test.fn;
+      event.test.fn = async () => {
+        try {
+          return await this.fixturePool.resolveParametersAndRun(fn);
+        } finally {
+          await this.fixturePool.teardownScope('test');
+        }
+      };
+    }
+  }
+}
+
+class Fixture {
+  constructor(pool, name, scope, fn) {
+    this.pool = pool;
+    this.name = name;
+    this.scope = scope;
+    this.fn = fn;
+    this.deps = fixtureParameterNames(this.fn);
+    this.usages = new Set();
+    this.value = null;
+  }
+
+  async setup() {
+    if (this._setup)
```

Comment:
We only call this from setupFixture(), so there should be no reentrancy? If there could be some reentrancy, shouldn't we call `await fixture.setup()` in line 141?

Reply:
Done. Should be all serialized.